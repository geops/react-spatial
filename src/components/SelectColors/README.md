#

This demonstrates the use of SelectColors.

```jsx
import React from 'react';
import SelectColors from 'react-spatial/components/SelectColors';

class SelectColorsExample extends React.Component {
  constructor(props) {
    super(props);
    this.colors = [
      { name: 'none', fill: [255, 255, 255, 0.01] },
      { name: 'black', fill: [0, 0, 0, 1] },
      { name: 'lightgrey', fill: [247, 244, 244, 1] },
      { name: 'darkblue', fill: [0, 61, 133, 1] },
      { name: 'lightgreen', fill: [123, 181, 76, 1] },
      { name: 'darkgreen', fill: [19, 111, 62, 1] },
    ];
    this.state = { color: this.colors[1] };
  }

  render() {
    return (
      <div className="tm-selectcolors-example">
        <SelectColors
          options={this.colors}
          value={this.state.color}
          onChange={color => {
            let index = this.colors.findIndex(x => x.name === color);
            let selectedColor = this.colors[index];
            this.setState({ color: selectedColor });
          }}
        />
        <div>You selected "{this.state.color.name}".</div>
      </div>
    );
  }
}

<SelectColorsExample />;
```
