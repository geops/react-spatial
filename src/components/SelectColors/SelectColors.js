import React from 'react';
import PropTypes from 'prop-types';

const propTypes = {
  /**
   * CSS class of the selectColors.
   */
  className: PropTypes.string,

  /**
   * Array of values to use in the selectColors element.
   */
  options: PropTypes.arrayOf(
    PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
      PropTypes.shape({
        value: PropTypes.oneOfType([
          PropTypes.string,
          PropTypes.number,
          PropTypes.arrayOf(
            PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
          ),
        ]).isRequired,
        label: PropTypes.string.isRequired,
      }),
    ]),
  ),

  /**
   * Function triggered on select's change event.
   * @param {Event} event The change event object.
   */
  onChange: PropTypes.func.isRequired,
};

const defaultProps = {
  options: undefined,
  className: 'tm-selectcolors',
};

/**
 * This component displays a simple selectColors HTML element.
 */
const SelectColors = function(props) {
  const { options } = props;
  const { className } = props;
  const { onChange } = props;
  if (!options) {
    return null;
  }

  return React.createElement(
    'select',
    {
      className,
      onChange(evt) {
        onChange(evt.target.value);
      },
    },
    (options || []).map(function(c) {
      return React.createElement('option', {
        key: c.name,
        value: c.name,
        style: { backgroundColor: `rgba(${c.fill})` },
      });
    }),
  );
};

SelectColors.propTypes = propTypes;
SelectColors.defaultProps = defaultProps;

export default SelectColors;
