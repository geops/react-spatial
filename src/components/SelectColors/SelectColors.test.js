import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import SelectColors from './SelectColors';

configure({ adapter: new Adapter() });

describe('SelectColors', () => {
  describe('when no properties are set', () => {
    let spy = null;

    beforeEach(() => {
      window.console.error = jest.fn().mockImplementation(() => {});
      spy = jest.spyOn(window.console, 'error');
    });

    afterEach(() => {
      spy.mockRestore();
      window.console.error.mockRestore();
    });

    test('matches snapshot', () => {
      const component = renderer.create(
        React.createElement(SelectColors, null),
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('when options are strings', () => {
    const options = ['foo', 'bar', 'baz'];

    test('matches snapshot', () => {
      const component = renderer.create(
        <SelectColors options={options} value="foo" onChange={() => {}} />,
      );

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('when options are objects with a value property.', () => {
    const options = [
      { name: 'none', fill: [255, 255, 255, 0.01] },
      { name: 'black', fill: [0, 0, 0, 1] },
      { name: 'blue', fill: [0, 0, 255, 1] },
    ];

    const onChange = () => {};

    test('matches snapshot', () => {
      const component = renderer.create(
        React.createElement(SelectColors, {
          options,
          value: options[1],
          onChange: () => {},
        }),
      );

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('executes onChange property function', () => {
      const onChangeMock = jest.fn(onChange);
      const wrapper = shallow(
        React.createElement(SelectColors, {
          options,
          value: options[2],
          onChange: onChangeMock,
        }),
      );
      expect(onChangeMock).toHaveBeenCalledTimes(0);
      const evt = { target: { name: 'none' } };
      wrapper.find('select').simulate('change', evt);
      expect(onChangeMock).toHaveBeenCalledTimes(1);
    });
  });
});
