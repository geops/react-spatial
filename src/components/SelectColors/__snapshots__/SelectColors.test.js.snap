// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectColors when no properties are set matches snapshot 1`] = `null`;

exports[`SelectColors when options are objects with a value property. matches snapshot 1`] = `
<select
  className="tm-selectcolors"
  onChange={[Function]}
>
  <option
    style={
      Object {
        "backgroundColor": "rgba(255,255,255,0.01)",
      }
    }
    value="none"
  />
  <option
    style={
      Object {
        "backgroundColor": "rgba(0,0,0,1)",
      }
    }
    value="black"
  />
  <option
    style={
      Object {
        "backgroundColor": "rgba(0,0,255,1)",
      }
    }
    value="blue"
  />
</select>
`;

exports[`SelectColors when options are strings matches snapshot 1`] = `
<select
  className="tm-selectcolors"
  onChange={[Function]}
>
  <option
    style={
      Object {
        "backgroundColor": "rgba(undefined)",
      }
    }
  />
  <option
    style={
      Object {
        "backgroundColor": "rgba(undefined)",
      }
    }
  />
  <option
    style={
      Object {
        "backgroundColor": "rgba(undefined)",
      }
    }
  />
</select>
`;
